### PR TITLE
Egg reproduction #8

### LINKED ISSUE #46: Sea lice eggs
Implement reproduction from sea lice eggs

Requires to know how many females have been fecundated, see #7

Currently done:
- [X] Implement main formula from Aldrin et al.
- [x] Testing
- [x] Integrate with mating events
- [x] Integrate with genomics data

### PR DESCRIPTION
**Pull request by erolm-a**
Closes #46 and #8 (for now).

As discussed offline, the mating code is currently deprecated in favour of a simpler logistic model as outline by Aldrin. However, pointers were left in case they were to be re-discussed.

### CODE REVIEW

File: refactor/src/Cage.py [hunk 1]
```diff
@@ -418,13 +426,36 @@ def do_mating_events(self):
                 df_list[fc].loc[df_list[fc].index.isin(dams),'mate_resistanceT1'] = \
                 df_list[fc].loc[df_list[fc].index.isin(sires),'resistanceT1'].values
         """
-        pass
 
-    def create_offspring(self):
+        # for now only consider a sensible number of lice that can mate
+        return min([10, self.lice_population["L5f"], self.lice_population["L5m"]])
```
Comment (SaraJakubiak) [inline refactor/src/Cage.py:431]: Why is 10 a sensible number?

File: refactor/src/Cage.py [hunk 2]
```diff
@@ -418,13 +426,36 @@ def do_mating_events(self):
...
         """
-        TODO
+        Get the number of new eggs
+        :param cur_month the current month (to compute the temperature)
+        :returns the number of eggs produced
         """
 
+        # See Aldrin et al. 2017, §2.2.6
+        # TODO: in Aldrin matings are not taken into account. This may change
+        female_population = self.lice_population['L5f']
```
Comment (SaraJakubiak) [inline refactor/src/Cage.py:442]: Can we keep consistency in usage of `'` and `"`?

File: refactor/src/Cage.py [hunk 3]
```diff
@@ -418,13 +426,36 @@ def do_mating_events(self):
...
+        age_distrib = self.get_stage_ages_distrib("L5f")
+        density_rate = female_population * age_distrib / self.num_fish
+        # density_rate = mated_females / self.num_fish
+        age_range = np.arange(1, len(age_distrib) + 1)
+
+        density_factor = np.ones_like(density_rate) - np.exp(-self.cfg.reproduction_density_dependence * density_rate)
+        ave_temp = self.farm.year_temperatures[cur_month - 1]
+        temperature_factor = self.cfg.delta_m10["L0"] * (10 / ave_temp) ** self.cfg.delta_p["L0"]
```
Comment (SaraJakubiak) [inline refactor/src/Cage.py:450]: Is 10 a constant from Aldrin here?

**Review by SaraJakubiak — CHANGES_REQUESTED** [review]

File: refactor/src/Cage.py (hunk 1 above)
Comment (erolm-a) [reply refactor/src/Cage.py:431]: I think I did this for testing, or at least I planned to but ended up not doing anything with it anyway. I'll revert this.

File: refactor/src/Cage.py (hunk 3 above)
Comment (erolm-a) [reply refactor/src/Cage.py:450]: delta_m10 means median development factor based on a fixed temperature of 10 degrees.

File: refactor/src/Cage.py (hunk 2 above)
Comment (erolm-a) [reply refactor/src/Cage.py]: Note: it would be good to adopt a linting strategy that covers this and other similar situations. See #14

**Review by SaraJakubiak — APPROVED** [review]

**Pull request merged**